Bag improvements

knothed opened the pull request.

knothed committed:
Fix wrong spelling

knothed committed:
Improve Bag

Add a "removeAll" method that replaces the mysterious "remove with negative value" functionality.

knothed committed:
Update Bag documentation

fredpi commented on Collections/Bag/Bag.swift:
```diff
@@ -66,15 +66,16 @@ struct Bag<ContainedElement: Hashable> {
-        } else {
-            contents.removeValue(forKey: member)
-        }
+        contents[member] = currentCount - occurrences
```
Please add handling for elements that are currently not contained. Otherwise a call would result in value of -1 for the given key.

knothed replied:
Fixed and improved.

knothed committed:
Fix type

knothed committed:
Improve Bag

Add an enum that distinguishes between removal modes.
Also, fix a bug.

Not tested yet.

fredpi commented on Collections/Bag/Bag.swift:
```diff
@@ -4,6 +4,12 @@
+    case any(Int)
+    case all
+}
+
```
Please include this as a private subtype to the Bag struct, not top-level

knothed replied:
Of course this is not possible, because an internal method cannot have a private type as a parameter.

fredpi commented on Collections/Bag/Bag.swift:
```diff
@@ -66,16 +72,26 @@ struct Bag<ContainedElement: Hashable> {
+            
+        case .all:
+            contents.removeValue(forKey: member)
+        }
```
Please document these changes

knothed replied:
Fixed.

knothed committed:
Move RemovalType into Bag

knothed committed:
Update Bag documentation regarding RemovalType
